backends/platform/atari/atari.cpp

```diff
@@ -0,0 +1,663 @@
+
+	Supexec(atari_ikbd_init);
+	_ikbd_initialized = true;
+
+	Supexec(asm_screen_falcon_save);
+	_video_initialized = true;
+
+	Setexc(VEC_PROCTERM, ikbd_and_video_restore);
+
+	BaseBackend::initBackend();
+}
+
+//! bit 0: rmb
+//! bit 1: lmb
+volatile uint8	g_atari_ikbd_mouse_buttons_state = 0;
+volatile int16	g_atari_ikbd_mouse_delta_x = 0;
+volatile int16	g_atari_ikbd_mouse_delta_y = 0;
+
+#define SCANCODES_SIZE 256
+volatile uint8	g_atari_ikbd_scancodes[SCANCODES_SIZE];
+uint16			g_atari_ikbd_scancodes_mask = SCANCODES_SIZE-1;
+volatile uint16	g_atari_ikbb_scancodes_head = 0;
+static uint16	g_atari_ikbb_scancodes_tail = 0;
+
+bool OSystem_Atari::pollEvent(Common::Event &event) {
+	{
+		static uint32 startMillis = getMillis();
+		static uint32 oldMillis = getMillis();
+		uint32 curMillis = getMillis();
+
+		uint32 diff = curMillis - oldMillis;
+		oldMillis = curMillis;
+
+		if (diff > 0) {
+			static float avgFpsSum;
+			static int avgFpsCount;
+
+			avgFpsSum += 1000.0f / diff;
+			avgFpsCount++;
+
+			if (curMillis - startMillis >= 1000) {
+				float avgFps = avgFpsSum / avgFpsCount;
+				debug("*** Average FPS in 1s: %f ***", avgFps);
+				startMillis = curMillis;
+				avgFpsSum = 0;
+				avgFpsCount = 0;
+			}
+		}
+	}
+
+	((DefaultTimerManager *)_timerManager)->checkTimers();
+	((AtariMixerManager *)_mixerManager)->update();
+
+	if ((g_atari_ikbd_mouse_buttons_state & 0x01) && !_oldRmbDown) {
+		event.type = Common::EVENT_RBUTTONDOWN;
+		event.mouse = _atariGraphicsManager->getMousePosition();
+		_oldRmbDown = true;
+		return true;
+	}
+
+	if (!(g_atari_ikbd_mouse_buttons_state & 0x01) && _oldRmbDown) {
+		event.type = Common::EVENT_RBUTTONUP;
+		event.mouse = _atariGraphicsManager->getMousePosition();
+		_oldRmbDown = false;
+		return true;
+	}
+
+	if ((g_atari_ikbd_mouse_buttons_state & 0x02) && !_oldLmbDown) {
+		event.type = Common::EVENT_LBUTTONDOWN;
+		event.mouse = _atariGraphicsManager->getMousePosition();
+		_oldLmbDown = true;
+		return true;
+	}
+
+	if (!(g_atari_ikbd_mouse_buttons_state & 0x02) && _oldLmbDown) {
+		event.type = Common::EVENT_LBUTTONUP;
+		event.mouse = _atariGraphicsManager->getMousePosition();
+		_oldLmbDown = false;
+		return true;
+	}
+
+	if (g_atari_ikbd_mouse_delta_x != 0 || g_atari_ikbd_mouse_delta_y != 0) {
+		const int deltaX = g_atari_ikbd_mouse_delta_x;
+		const int deltaY = g_atari_ikbd_mouse_delta_y;
+
+		g_atari_ikbd_mouse_delta_x = g_atari_ikbd_mouse_delta_y = 0;
+
+		_atariGraphicsManager->updateMousePosition(deltaX, deltaY);
+
+		event.type = Common::EVENT_MOUSEMOVE;
+		event.mouse = _atariGraphicsManager->getMousePosition();
+		event.relMouse = Common::Point(deltaX, deltaY);
+		return true;
+	}
+
+	if (g_atari_ikbb_scancodes_head != g_atari_ikbb_scancodes_tail) {
+		byte scancode = g_atari_ikbd_scancodes[g_atari_ikbb_scancodes_tail++];
+		g_atari_ikbb_scancodes_tail &= SCANCODES_SIZE-1;
+
+		bool pressed = !(scancode & 0x80);
+		scancode &= 0x7f;
+
+		if (scancode == 0x1d)
```

Comment:
What about using a `switch()` statement here?

Reply:
I don't think that is possible - all the keys can be pressed simultaneously.